core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -0,0 +1,939 @@
+
+  /* offsets cache */
+  private val offsetsCache = new Pool[GroupTopicPartition, OffsetAndMetadata]
+
+  /* group metadata cache */
+  private val groupsCache = new Pool[String, GroupMetadata]
+
+  /* partitions of consumer groups that are being loaded, its lock should be always called BEFORE the group lock if needed */
+  private val loadingPartitions: mutable.Set[Int] = mutable.Set()
+
+  /* partitions of consumer groups that are assigned, using the same loading partition lock */
+  private val ownedPartitions: mutable.Set[Int] = mutable.Set()
+
+  /* lock for expiring stale offsets, it should be always called BEFORE the group lock if needed */
+  private val offsetExpireLock = new ReentrantReadWriteLock()
+
+  /* lock for removing offsets of a range partition, it should be always called BEFORE the group lock if needed */
+  private val offsetRemoveLock = new ReentrantReadWriteLock()
+
+  /* shutting down flag */
+  private val shuttingDown = new AtomicBoolean(false)
+
+  /* number of partitions for the consumer metadata topic */
+  private val groupMetadataTopicPartitionCount = getOffsetsTopicPartitionCount
+
+  this.logIdent = "[Group Metadata Manager on Broker " + brokerId + "]: "
+
+  scheduler.schedule(name = "delete-expired-consumer-offsets",
+    fun = deleteExpiredOffsets,
+    period = config.offsetsRetentionCheckIntervalMs,
+    unit = TimeUnit.MILLISECONDS)
+
+  newGauge("NumOffsets",
+    new Gauge[Int] {
+      def value = offsetsCache.size
+    }
+  )
+
+  newGauge("NumGroups",
+    new Gauge[Int] {
+      def value = groupsCache.size
+    }
+  )
+
+  def partitionFor(groupId: String): Int = Utils.abs(groupId.hashCode) % groupMetadataTopicPartitionCount
+
+  def isGroupLocal(groupId: String): Boolean = loadingPartitions synchronized ownedPartitions.contains(partitionFor(groupId))
+
+  def isGroupLoading(groupId: String): Boolean = loadingPartitions synchronized loadingPartitions.contains(partitionFor(groupId))
+
+  /**
+   * Get the group associated with the given groupId, or null if not found
+   */
+  def getGroup(groupId: String): GroupMetadata = {
+      groupsCache.get(groupId)
+  }
+
+  /**
+   * Add a group or get the group associated with the given groupId if it already exists
+   */
+  def addGroup(groupId: String, protocolType: String): GroupMetadata = {
+    addGroup(groupId, new GroupMetadata(groupId, protocolType))
+  }
+
+  private def addGroup(groupId: String, group: GroupMetadata): GroupMetadata = {
+    groupsCache.putIfNotExists(groupId, group)
+    group
+  }
+
+  /**
+   * Remove all metadata associated with the group, note this function needs to be
+   * called inside the group lock
+   * @param group
+   */
+  def removeGroup(group: GroupMetadata) {
+    // first mark the group as dead
+    group.transitionTo(Dead)
+
+    if (groupsCache.remove(group.groupId) != group)
+      throw new IllegalArgumentException("Cannot remove group " + group.groupId + " since it has been replaced.")
```

Comment:
This case would happen if two threads hit removeGroup simultaneously, right? If there's nothing preventing that case, would we be better returning rather than raising an exception?